typed-racket-lib/typed-racket/base-env/unit-prims.rkt

```diff
@@ -0,0 +1,767 @@
+             #:attr link-units? #t
+             #:with untyped-stx
+             #`(link uid-inits ... uid-last)
+             #:attr unit-ids (syntax->list #'(uid-inits ... uid-last))
+             #:with last-id 
+             (local-expand #'uid-last (syntax-local-context) (kernel-form-identifier-list)))))
+
+;; Note: This may not correctly handle all use cases of
+;; define-values/invoke-unit/infer
+;; inferred imports and exports are handled in the following way
+;; - the exports of ALL units being linked are added to the export list
+;;   to be registered in tc-toplevel
+;; - imports are the set-difference of the union of all imports and the
+;;   union of all exports
+;; This seems to correctly handle bot recursive and non-recursive
+;; linking patterns
+(define-syntax (define-values/invoke-unit/infer stx)
+  (syntax-parse stx
+    [dviui:define/invoke/infer-form
+     #`(begin
+         #,(internal (quasisyntax/loc stx
+                       (define-values/invoke-unit-internal
+                         (#,@(map imports/members (attribute dviui.inferred-imports)))
+                         (#,@(process-dv-exports (attribute dviui.inferred-exports))))))
+         #,(ignore (quasisyntax/loc stx dviui.untyped-stx)))]))
+
+;; invoke-unit macro
+(begin-for-syntax
+  (define-splicing-syntax-class invoke-imports
+    #:literals (import)
+    (pattern (~seq)
+             #:attr untyped-import #'()
+             #:with imports #'())
+    (pattern (import sig:id ...)
+             #:attr untyped-import #'((import sig ...))
+             #:with imports #'((quote-syntax sig) ...))))
+
+
+;; need to do extra work to make this work with the existing
+;; invoke-unit machinery
+;; need to also add the local-expanded unit-ids to the table ..
+(define-syntax (invoke-unit/infer stx)
+  (syntax-parse stx 
+    [(_ us:unit-spec)
+     (define imports (map
+                      (lambda (sig) (replace-context stx sig))
+                      (infer-imports (attribute us.unit-ids))))
+     (define unit-expr-id #'us.last-id)
+     (ignore
+      (tr:unit:invoke
+       (quasisyntax/loc stx
+         (#%expression
+          (begin
+            (void (quote-syntax #,unit-expr-id))
+            (void #,@(map (lambda (id) #`(quote-syntax #,id)) imports))
+            (untyped-invoke-unit/infer us.untyped-stx))))))]))
+
+(define-syntax (invoke-unit stx)
+  (syntax-parse stx
+    [(invoke-unit unit-expr imports:invoke-imports)
+     (ignore
+      (tr:unit:invoke
+       (quasisyntax/loc stx
+         (#%expression
+          (begin
+            (void)
+            (void #,@#'imports.imports)
+            (untyped-invoke-unit
+             #,(tr:unit:invoke:expr-property
+                #'unit-expr
+                #t)
+             #,@(attribute imports.untyped-import)))))))]))
+
+;; This table implementation is going to break when only/except are allowed in
+;; typed units, the indexing strategy won't work in that case
```

Comment:
I don't understand this comment. Are you saying with those forms *none* of our approaches will work or some of them will?